org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/LSPDocumentLinkPresentationReconcilingStrategy.java

```diff
@@ -0,0 +1,229 @@
+
+		@Override
+		public void inputDocumentAboutToBeChanged(@Nullable final IDocument oldDoc, @Nullable final IDocument newDoc) {
+			if (oldDoc == null)
+				return;
+
+			final var viewer = LSPDocumentLinkPresentationReconcilingStrategy.this.viewer;
+			if (viewer != null) {
+				viewer.removeTextListener(this);
+			}
+		}
+
+		@Override
+		public void inputDocumentChanged(@Nullable final IDocument oldDoc, @Nullable final IDocument newDoc) {
+			if (newDoc == null) {
+				return;
+			}
+
+			final var viewer = LSPDocumentLinkPresentationReconcilingStrategy.this.viewer;
+			if (viewer == null)
+				return;
+			viewer.addTextListener(this);
+		}
+
+		@Override
+		public void textChanged(final @Nullable TextEvent event) {
+			if (event == null || !event.getViewerRedrawState()) {
+				return;
+			}
+
+			underline();
+
+		}
+
+	}
+
+	private void underline() {
+		final var viewer = LSPDocumentLinkPresentationReconcilingStrategy.this.viewer;
+		if (viewer == null)
+			return;
+
+		final IDocument document = viewer.getDocument();
+		if (document == null) {
+			return;
+		}
+
+		URI uri = LSPEclipseUtils.toUri(document);
+		if (uri == null) {
+			return;
+		}
+		cancel();
+		final DocumentLinkParams params = new DocumentLinkParams(new TextDocumentIdentifier(uri.toString()));
+		request = LanguageServiceAccessor
+				.getLanguageServers(document, capabilities -> capabilities.getDocumentLinkProvider() != null)
+				.thenAcceptAsync(languageServers -> CompletableFuture.allOf(languageServers.stream()
+						.map(languageServer -> languageServer.getTextDocumentService().documentLink(params))
+						.map(request -> request.thenAcceptAsync(links -> {
+							final Control control = viewer.getTextWidget();
+							if (control != null) {
+								control.getDisplay().asyncExec(() -> underline(links));
+							}
+						})).toArray(CompletableFuture[]::new)));
+	}
+
+	private void underline(List<DocumentLink> links) {
+		for (DocumentLink link : links) {
+			if (viewer != null) {
+				try {
+					IDocument document = viewer.getDocument();
+
+					// Compute link region
+					int start = LSPEclipseUtils.toOffset(link.getRange().getStart(), document);
+					int end = LSPEclipseUtils.toOffset(link.getRange().getEnd(), document);
+					int length = end - start;
+					IRegion linkRegion = new Region(start, length);
+
+					// Update existing style range with underline or create a new style range with
+					// underline
+					StyleRange styleRange = null;
+					StyleRange[] styleRanges = viewer.getTextWidget().getStyleRanges(start, length);
+					if (styleRanges != null) {
+						// It exists some styles for the range of document link, update just the
+						// underline style.
+						for (StyleRange s : styleRanges) {
+							s.underline = true;
+						}
+						TextPresentation presentation = new TextPresentation(linkRegion, 100);
+						presentation.replaceStyleRanges(styleRanges);

```

Comment:
Shouldn't it be `mergeStyleRanges` so the existing other styles (eg colors) should be kept?

Reply:
@angelozerr , I think the change would be ready to be merged, but there is still this pending question on why use replace and not `mergeStyleRanges`. If that is clear, I would merge it ;)

Reply:
TextPresentation is created only here so add, replace, merge will do the same thing I think.

There are several problems with this PR:

 * the request is not canceled properly on language server side.
 * the result of request is not canceled. You can have a lot of error logs with BadLocation (like Outline or other operations) and you will update the UI to many times althouth it should be updated just with last changes.

Those problems can be improved in https://github.com/eclipse/lsp4e/issues/236

It create an extra background Thread with reconciler although is is never used. This issue can be fixed like https://github.com/eclipse/lsp4e/pull/241